lib/jxl/render_pipeline/simple_render_pipeline.cc

```diff
@@ -6,16 +6,146 @@
 #include "lib/jxl/render_pipeline/simple_render_pipeline.h"
 
 namespace jxl {
-void SimpleRenderPipeline::PrepareForThreads(size_t num) {
-  // TODO(veluca): actually allocate input buffers.
+void SimpleRenderPipeline::PrepareForThreadsInternal(size_t num) {
+  for (size_t c = 0; c < channel_shifts_[0].size(); c++) {
+    channel_data_.push_back(ImageF(DivCeil(frame_dimensions_.xsize_padded,
+                                           1 << channel_shifts_[0][c].first) +
+                                       kRenderPipelineXOffset * 2,
+                                   DivCeil(frame_dimensions_.ysize_padded,
+                                           1 << channel_shifts_[0][c].second) +
+                                       kRenderPipelineXOffset * 2));
```

Comment:
These kind of computations of dimensions (and rects, like in PrepareBuffers) are probably better factored out into auxiliary functions. That will help to make this code easier to read and thus more 'obviously correct' :)

(also such auxiliary functions will probably be useful also in the non-simple render pipeline)

Reply:
They will definitely *not* be useful in the non-simple render pipeline :)